Content.Server/Xenoarchaeology/Artifact/XenoArtifactSystem.ProcGen.cs

```diff
@@ -0,0 +1,176 @@
+using System.Linq;
+using Content.Shared.Random.Helpers;
+using Content.Shared.Xenoarchaeology.Artifact.Components;
+using Content.Shared.Xenoarchaeology.Artifact.Prototypes;
+using Robust.Shared.Random;
+
+namespace Content.Server.Xenoarchaeology.Artifact;
+
+public sealed partial class XenoArtifactSystem
+{
+    private readonly List<XenoArchTriggerPrototype> _triggerPool = new();
+
+    private void GenerateArtifactStructure(Entity<XenoArtifactComponent> ent)
+    {
+        var nodeCount = ent.Comp.NodeCount.Next(RobustRandom);
+        CreateTriggerPool(ent, ref nodeCount);
+        ResizeNodeGraph(ent, nodeCount);
+        while (nodeCount > 0)
+        {
+            GenerateArtifactSegment(ent, ref nodeCount);
+        }
+
+        RebuildXenoArtifactMetaData((ent, ent));
+    }
+
+    private void CreateTriggerPool(Entity<XenoArtifactComponent> ent, ref int size)
+    {
+        _triggerPool.Clear();
+        var weightsProto = PrototypeManager.Index(ent.Comp.TriggerWeights);
+        var weights = new Dictionary<string, float>(weightsProto.Weights);
+
+        while (_triggerPool.Count < size)
+        {
+            // OOPS! We ran out of triggers.
+            if (weights.Count == 0)
```

Comment:
This one is wierd. Mechanism is making 1 trigger per node, but tier1+ nodes include previous nodes triggers to activate.
But to change node count when having not enough trigger - this is highly undexpected behaviour...

Reply:
After all, this is part of overall design decision. It might be changed when we addition to original design will be ready - node effect affixes will require re-doing this stuff, so this might be fine for now